alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs

```diff
@@ -0,0 +1,466 @@
+
+  txins @era txb /= Set.empty ?! InputSetEmptyUTxO
+
+  feesOK pp tx utxo
+
+  let minimumFee = minfee pp tx
+      txFee = getField @"txfee" txb
+  minimumFee <= txFee ?! FeeTooSmallUTxO minimumFee txFee
+
+  eval (txins @era txb ⊆ dom utxo)
+    ?! BadInputsUTxO (eval ((txins @era txb) ➖ (dom utxo)))
+
+  ni <- liftSTS $ asks networkId
+  let addrsWrongNetwork =
+        filter
+          (\a -> getNetwork a /= ni)
+          (fmap (getField @"address") $ toList $ getField @"outputs" txb)
+  null addrsWrongNetwork ?! WrongNetwork ni (Set.fromList addrsWrongNetwork)
+  let wdrlsWrongNetwork =
+        filter
+          (\a -> getRwdNetwork a /= ni)
+          (Map.keys . unWdrl . getField @"wdrls" $ txb)
+  null wdrlsWrongNetwork
+    ?! WrongNetworkWithdrawal
+      ni
+      (Set.fromList wdrlsWrongNetwork)
+
+  let consumed_ = consumed pp utxo txb
+      produced_ = Shelley.produced @era pp stakepools txb
+  consumed_ == produced_ ?! ValueNotConservedUTxO consumed_ produced_
+
+  -- process Protocol Parameter Update Proposals -- NOT SURE WHAT IS GOING On HERE
+  _ppup' <-
+    trans @(Core.EraRule "PPUP" era) $
+      TRC (PPUPEnv slot pp genDelegs, ppup, txUpdates txb)
+
+  -- Check that the mint field does not try to mint ADA. This is equivalent to
+  -- the check `adaPolicy ∉ supp mint tx` in the spec.
+  Val.coin (getField @"mint" txb) == Val.zero ?! TriesToForgeADA
+
+  let outputs = Map.elems $ unUTxO (txouts @era txb)
+      minUTxOValue = getField @"_minUTxOValue" pp
+      outputsTooSmall =
+        filter
+          ( \out ->
+              let v = getField @"value" out
+               in not $
+                    Val.pointwise
+                      (>=)
+                      v
+                      (Val.inject $ scaledMinDeposit v minUTxOValue)
+          )
+          outputs
+  null outputsTooSmall ?! OutputTooSmallUTxO outputsTooSmall
+
+  let outputsTooBig =
+        filter
+          ( \out ->
+              let v = getField @"value" out
+               in (BSL.length . serialize) v > 4000
+              -- TODO this is arbitrary, THERE IS SUPPOSEDLY A NEW
+              -- PParams files that holds this what is it? --TODO fix this
+          )
+          outputs
+  null outputsTooBig ?! OutputTooBigUTxO outputsTooBig
+
+  -- Bootstrap (i.e. Byron) addresses have variable sized attributes in them.
+  -- It is important to limit their overall size.
+  let outputsAttrsTooBig =
+        filter
+          ( \out -> case getField @"address" out of
+              AddrBootstrap addr -> bootstrapAddressAttrsSize addr > 64
+              _ -> False
+          )
+          outputs
+  null outputsAttrsTooBig ?! OutputBootAddrAttrsTooBig outputsAttrsTooBig
+
+  let maxTxSize_ = fromIntegral (getField @"_maxTxSize" pp)
+      txSize_ = txsize tx
+  txSize_ <= maxTxSize_ ?! MaxTxSizeUTxO txSize_ maxTxSize_
+
+  let maxTxEx = getField @"_maxTxExUnits" pp
+  txExunits txb <= maxTxEx ?! ExUnitsTooSmallUTxO maxTxEx (txExunits txb)
```

Comment:
this will have to sum the `exunits` in the `txrdmrs` also (the same as in the minfee calc)